bot/utils/scheduling.py

```diff
@@ -51,20 +51,32 @@ def schedule_task(self, task_id: t.Hashable, task_data: t.Any) -> None:
         self._scheduled_tasks[task_id] = task
         log.debug(f"{self.cog_name}: scheduled task #{task_id} {id(task)}.")
 
-    def cancel_task(self, task_id: t.Hashable) -> None:
-        """Unschedule the task identified by `task_id`."""
+    def cancel_task(self, task_id: t.Hashable, ignore_missing: bool = False) -> None:
+        """
+        Unschedule the task identified by `task_id`.
+
+        If `ignore_missing` is True, a warning will not be sent if a task isn't found.
+        """
         log.trace(f"{self.cog_name}: cancelling task #{task_id}...")
         task = self._scheduled_tasks.get(task_id)
 
         if not task:
-            log.warning(f"{self.cog_name}: failed to unschedule {task_id} (no task found).")
+            if not ignore_missing:
+                log.warning(f"{self.cog_name}: failed to unschedule {task_id} (no task found).")
             return
 
-        task.cancel()
         del self._scheduled_tasks[task_id]
+        task.cancel()
 
         log.debug(f"{self.cog_name}: unscheduled task #{task_id} {id(task)}.")
 
+    def cancel_all(self) -> None:
+        """Unschedule all known tasks."""
+        log.debug(f"{self.cog_name}: unscheduling all tasks")
+
+        for task_id in self._scheduled_tasks.copy():
+            self.cancel_task(task_id, ignore_missing=True)
+
```

Comment:
After cancelling the tasks, you may also want to consider propagating an exception for any task that couldn't be cancelled; it can help to make subtle errors more apparent. We use this in the implementation for `asyncio.run()`. For an example of how this could be done, see https://github.com/python/cpython/blob/909f4a30093f74d409711e564f93a43167ca0919/Lib/asyncio/runners.py#L62-L73.

Reply:
@aeros All tasks have a [done callback](https://github.com/python-discord/bot/blob/master/bot/utils/scheduling.py#L80) which retrieves exceptions from tasks. When a task is cancelled, it is also considered done. Did I cover all bases here or am I still missing some scenario?

Reply:
Oh I think what you're getting at is when `task.cancel()` is called but `task.cancelled()` still returns False, meaning it wasn't actually cancelled.

Reply:
@MarkKoz 
> Oh I think what you're getting at is when task.cancel() is called but task.cancelled() still returns False, meaning it wasn't actually cancelled

Yep, that's what I was referring to. It's specifically when task.cancel() is called, `task.cancelled()` returns false, and `task.exception()` contains an exception of some form. Your `_task_done_callback()` does cover this for the most part (I hadn't looked over it), but indefinitely running tasks may still be a concern with that implementation.

> So what if I see it's not cancelled but really all it needed was a little bit more time?

For this situation, you could consider using a timeout at the end with `asyncio.wait_for()`, to give your tasks N seconds at most to cancel. This would prevent any hanging tasks from indefinitely locking up the shutdown process, if that might be an issue. In your situation, you'd likely want to use it where you use `cancel_all()` in `cog_unload()`. For example:

```py
    def cog_unload(self) -> None:
        """Cancel the init task and scheduled tasks when the cog unloads."""
        log.trace("Cog unload: cancelling the init_cog task")
        self.init_task.cancel()

        log.trace("Cog unload: cancelling the channel queue tasks")
        for task in self.queue_tasks:
            task.cancel()
        
        # You'll likely have to determine an optimal timeout, but 300 secs (10 mins)
        # could be a reasonable default.
        try:
            await asyncio.wait_for(self.cancel_all(), timeout=300)
        except asyncio.TimeoutError:
            log.error("could not cancel all tasks: timeout occurred")
```

Reply:
`cancel_task()` essentially just calls `task.cancel()` for each task. I thought `task.cancel()` didn't block, so how would a timeout help?

Reply:
Oh and I forgot that a task is also considered done when an exception is raised inside it, so presumably that means the done callback would also trigger if an exception was raised when there was an attempt to cancel a task? In that case, yeah, I do believe you're right that the callback is covering it.

Reply:
> I thought task.cancel() didn't block, so how would a timeout help?

I was just about to correct the above, sorry: you'd have to modify `cancel_all()` slightly (or add it in `cog_unload`) to wait for `task.done()` to be true for each task that still isn't cancelled after attempting to call `task.cancel()`.

Reply:
But the above is only needed if you think indefinitely running tasks might be an issue (which you'll likely have a better idea than I do, based on what the tasks are doing). It might not be a realistic concern in your case.

Reply:
No, not indefinite, but there can be long sleeps (30 minutes). In any case, I think it's safe to leave it alone for now but I will keep this in mind if it turns out to be a problem. Thanks.

Reply:
> Oh and I forgot that a task is also considered done when an exception is raised inside it, so presumably that means the done callback would also trigger if an exception was raised when there was an attempt to cancel a task?

Yep, it's only an issue when the exception is in the process of propagating/being handled, but hasn't been cancelled yet. If it finished raising, the task is considered done, so it should trigger the done callback when the exception was raised.

As mentioned above, I'm not certain the above will be a realistic concern in your case. As the one who wrote it, you'll likely have a better understanding of whether or not `_task_done_callback()` will be adequate based on what the tasks are actually doing. If indefinitely hanging tasks are an issue later down the road, I'd recommend revisiting it though and implementing some form of timeout.